text/0112-compress-state-response-message-in-state-sync.md

```diff
@@ -0,0 +1,70 @@
+# RFC-0112: Compress the State Response Message in State Sync
+
+|                 |                                                                                             |
+| --------------- | ------------------------------------------------------------------------------------------- |
+| **Start Date**  | 14 August 2024                                                                              |
+| **Description** | Compress the state response message to reduce the data transfer during the state syncing    |
+| **Authors**     | Liu-Cheng Xu                                                                                |
+
+## Summary
+
+This RFC proposes compressing the state response message during the state syncing process to reduce the amount of data transferred.
+
+## Motivation
```

Comment:
Is the motivation to send less data or to improve state sync performance? Based on what you're saying, I think that implementing a persistence feature would be enough to solve this issue on slower network connections, as it allows you to request subsequent portions of the state incrementally.

Perhaps a better way to solve it is a combination of a persistence layer to store the state you are retrieving and to be able to stop/restart your node, and perhaps changing the request to ask for a certain amount of state.

Currently, the state request is like [this](https://github.com/paritytech/polkadot-sdk/blob/fa52407856c11ee138a32f2ba744f774fac984d5/substrate/client/network/sync/src/schema/api.v1.proto#L73-L81)

```protobuf
message StateRequest {
    // Block header hash.
    bytes block = 1;
    // Start from this key.
    // Multiple keys used for nested state start.
    repeated bytes start = 2; // optional
    // if 'true' indicates that response should contain raw key-values, rather than proof.
    bool no_proof = 3;
}
```

What I'm suggesting is to add a new field with a limit:

```protobuf
message StateRequest {
    // Block header hash.
    bytes block = 1;
    // Start from this key.
    // Multiple keys used for nested state start.
    repeated bytes start = 2; // optional
    // if 'true' indicates that response should contain raw key-values, rather than proof.
    bool no_proof = 3;
    // amount of keys we want to receive in the response
    uint32 max_keys = 4;
}
```

I know this could be configured as a different proposal, but I think it could be a better way to solve what you are saying without relying on compression algorithms. In the worst case, you could use a combination of both, asking for a partial state + compressing the response.

The trade-off might be that the number of requests could increase. wdyt?

Reply:
Thanks for your feedback. There is already a persistent state sync tracking issue https://github.com/paritytech/polkadot-sdk/issues/4 that I'm actively working on. I initiated this RFC because this is a low-hanging fruit to reduce the bandwidth that can lead to immediate improvements in state sync performance, while persistent state sync requires much more effort and complexity (_may take too long to wrap it up_). As you can see from the patch, a few change lines will significantly improve the bandwidth consumption. The proposed changes in the RFC are minimal yet impactful in terms of bandwidth savings.

In contrast, your suggestion to add a `max_keys` field, while valid, does not contribute to reducing any bandwidth. It may even lead to increased request frequency, which could negate some of the performance gains we seek.